### ISSUE TITLE
Refactor and separate client formats

### ISSUE THREAD

**Issue by dylanede** [tomaka/glium#451]
Currently, `ClientFormat` encodes both the layout of channels and their datatypes (the layout is implicitly always RGBA or a truncated version of that). In order to eventually fully support all of the formats supported by OpenGL without leading to a combinatorial explosion, `ClientFormat` should be split into two enums - one for channel layout (e.g. `RGBA`, `BGR`, `RG`) and one for unit datatype (e.g. `U8`, `I8`, `U16`, `U_3_3_2`). Not all  combinations are valid, but these can be detected at run time.

**Comment by tomaka** [comment]
I don't think this is a good idea.

As I said on IRC, with modern OpenGL you're no longer manipulating "red", "green", "blue" and "alpha", but just components.
Textures can be used to store more than just colors, and having concepts like "RGBA" in glium is confusing.

However we can provide a way for the user to exchange components while uploading.

**Comment by kvark** [comment]
I don't think this is a bad idea, we have it separated (in another project). You can store whatever you want in textures, but the shaders still access it as RGBA, so the components semantics is built into GL.

**Comment by dylanede** [comment]
Don't even think of them as colours. I don't. They're just R, G, B, and A, our four faithful channels. Rename them to C0, C1, C2 and C3 if you want. You still need to separate channel ordering from channel data types.
